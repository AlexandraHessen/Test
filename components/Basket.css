.TableBasket{
    width: 100%;
    border-collapse: collapse;
    margin-bottom: 25px;
}

.TableBasket tr td{
    padding: 20px 0;
    width: 10%;
}

.TableBasket .TableHeader{
    text-align: left;
    padding-bottom: 10px;
}

.TableBasket .TableHeader th{
    padding: 10px 0 10px;
    color: #969696;
}

.Total{
    text-align: right;
}

.SumProductsInBasket{
    font-size: 22px;
    font-weight: bold;
    margin-bottom: 20px;
}

.ActionsBasket{
    clear: both;
    text-align: right;
}

.BasketButtons{
    border: solid 2px #8ccb68;
    font-weight: bold;
    color:  #363636;
    padding: 8px 13px;
    border-radius: 40px;
    cursor: pointer;
    position: relative;
    background-color: #fff;
    font-family: 'Rubik', sans-serif;
    margin: 0 0 25px 20px;

}

.priceTh{
    text-align: center;
}