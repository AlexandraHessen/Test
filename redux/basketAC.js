// ---------------- ActionCreator ----------------//
const ADD_PRODUCT = "ADD_PRODUCT"
const DEL_PRODUCT = "DEL_PRODUCT"
const CLEAR_BASKET = "CLEAR_BASKET" //пользователь сам очищает корзину при желании

//название action type

//функция для dispatch (из файла CounterButton) которая возвращает хэш с action type 
// называется ActionCreator

const add_product = function(productId, objProductInfo) {
    return {
      type: ADD_PRODUCT,
      productId: productId,
      objProductInfo: objProductInfo
    };
  }
  
  const del_product = function(productId) {
    return {
      type: DEL_PRODUCT,
      productId: productId,
    };
  }

  const clear_basket = function() {
    return {
      type: CLEAR_BASKET
    };
  }

export {
    add_product, ADD_PRODUCT,
    del_product, DEL_PRODUCT,
    clear_basket, CLEAR_BASKET,
  }