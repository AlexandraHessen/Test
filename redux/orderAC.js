// ---------------- ActionCreator ----------------//
const ADD_ORDER = "ADD_ORDER"

//название action type

//функция для dispatch (из файла CounterButton) которая возвращает хэш с action type 
// называется ActionCreator

const add_order = function(objCustomerInfo, objOrderInfo) {
    return {
      type: ADD_ORDER,
      objCustomerInfo: objCustomerInfo,
      objOrderInfo: objOrderInfo
    };
  }

export {
    add_order, ADD_ORDER,
  }