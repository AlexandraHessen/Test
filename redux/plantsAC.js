// ---------------- ActionCreator ----------------//
const PLANTS_LOADING='PLANTS_LOADING';
const PLANTS_ERROR='PLANTS_ERROR';
const PLANTS_SET='PLANTS_SET';
//название action type

//функция для dispatch (из файла CounterButton) которая возвращает хэш с action type 
// называется ActionCreator
const plantsLoadingAC=function() {
  return {
    type: PLANTS_LOADING,
  };
}

const plantsErrorAC=function() {
  return {
    type: PLANTS_ERROR,
  };
}

const plantsSetAC=function(plants) {
  return {
    type: PLANTS_SET,
    plants:plants,
  };
}

export {
  plantsLoadingAC,PLANTS_LOADING,
  plantsErrorAC,PLANTS_ERROR,
  plantsSetAC,PLANTS_SET,
}
