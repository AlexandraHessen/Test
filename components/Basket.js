"use strict";

import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

import {connect} from 'react-redux';
import {clear_basket} from '../redux/basketAC';

import RowProductBasket from './RowProductBasket '
import Order from './Order';
import './Basket.css'

class Basket extends React.PureComponent{

    static propTypes = {
        basket: PropTypes.object.isRequired // пришло из Redux для проверки есть ли что-то в корзине
                                            // в ответ ничего не передаем 
                                            // ключ productsInBasket
    };
    //    получим значение нужного счётчика
    // let counterValue=this.props.counters.cnts[this.props.counterid];
    // let counterValue=this.props.basket.productsInBasket;
    state={
      checkout: false //нажата кнопка Оформить заказ
    }

    clearBasket=()=>{
      this.props.dispatch( clear_basket() )
    }

    checkout=()=>{
      this.setState({checkout: true})
    }

    render(){
        let sumProductsInBasket=0;
        let productsInBasketCode=[];
        //НУЖНО ДЕЛАТЬ МАССИВ

        for (let prod in this.props.basket.productsInBasket) {
    //  console.log(prod) //12 =ID
    //  console.log(this.state.cart.products)//obj из всех товаров
    //  console.log(this.props.basket.productsInBasket[prod].code) //12
    //  console.log(this.state.cart.products[prod])//данные этого товара {id: 11, name: 'Морковь Вита Лонга', brand: 'Гавриш', category: 'Семена овощей', price: 0.65, …}
            productsInBasketCode.push(<RowProductBasket 
                key = {this.props.basket.productsInBasket[prod].code} 
                info = {this.props.basket.productsInBasket[prod]} />)
                sumProductsInBasket+=this.props.basket.productsInBasket[prod].price
          };
          
        return(
          <div>
            <table className = 'TableBasket'>
              <tbody>
                <tr className = 'TableHeader'>
                  <th>Товар</th>
                  <th></th>
                  <th className='priceTh'>Цена</th>
                  {/* <th>Количество</th>
                  <th>Итого</th> */}
                  <th></th>
                </tr>
                {productsInBasketCode}
              </tbody>
            </table>
            <div className='Total'>
              Итого:  
              <div className='SumProductsInBasket'>{sumProductsInBasket} руб.</div>
            </div>
            <div className='ActionsBasket'>
              <input type="button" value="Очистить корзину" className="BasketButtons" onClick = {this.clearBasket}></input>
              <input type="button" value="Оформить заказ" className="BasketButtons" onClick = {this.checkout}></input>
            </div>
            {this.state.checkout && <Order />}
          </div>
            
        )
    }
    

}

const mapStateToProps = function (state) {
    return {
      // весь раздел Redux state под именем basket будет доступен
      // данному компоненту как this.props.basket
      basket: state.basket,
    };
  };
  
  export default connect(mapStateToProps)(Basket);
  // экспортируем не этот класс CounterButton а уже обвернутый