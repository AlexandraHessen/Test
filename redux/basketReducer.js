// ---------------- САМА ФУНКЦИЯ REDUСER  ----------------//
// REDUCER не делает запрос в Ajax, логика должна быть в компоненте
import { ADD_PRODUCT, DEL_PRODUCT, CLEAR_BASKET } from './basketAC'; //название action type


const initState={ // начальный state
    productsInBasket: {}
    //КЛЮЧИ productId, objProductInfo
    // 1: {code: 1, name: 'Замиокулькас'}

//   status: 0, // 0 - ничего не началось, 1 - идёт загрузка, 2 - была ошибка, 3 - данные загружены
//   data: null, //даные которые мы загрузили по сети

}

function basketReducer(state=initState,action) {
  switch (action.type) {

    case ADD_PRODUCT: {
      console.log('action:',action);
      console.log('state до обработки редьюсером:',state);
      let newState={...state, //старый state
        productsInBasket:{...state.productsInBasket, //заменить старые продукты на теже +
          // ПРИЛЕТИТ какой продукт (productid) изменить и насколько(addvalue)
          [action.productId]:action.objProductInfo //+
          //[action.counterid] - В нужном продукте 
          // state.productsInBasket[action.productid] - его старое значение 
          // action.addvalue - изменить на прилетевшее
        }
      };
      console.log('state после обработки редьюсером:',newState);
      return newState;
    }

    case DEL_PRODUCT: {
      let newState = {...state};
      delete newState.productsInBasket[action.productId];
      return newState;
    }

    case CLEAR_BASKET:{
      let newState = {...state};
      newState.productsInBasket={};
      return newState;
    }

    // case PLANTS_ERROR: {
    //   let newState={
    //     status:2,
    //     data:null,
    //   };
    //   return newState;
    // }

    // case PLANTS_SET: {
    //   let newState={
    //     status:3,
    //     data:action.plants, //загруженные данные
    //   };
    //   return newState;
    // }
    
//обязатнльно пишем на тот случай если пришел action который не относится к этому reducery
    default:
      return state;
  }
}

export default basketReducer;
