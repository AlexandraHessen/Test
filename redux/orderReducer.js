// ---------------- САМА ФУНКЦИЯ REDUСER  ----------------//
// REDUCER не делает запрос в Ajax, логика должна быть в компоненте
import { ADD_ORDER } from './orderAC'; //название action type

const initState={ // начальный state
    orders: []
}

function orderReducer(state=initState,action) {
  switch (action.type) {
    case ADD_ORDER: {
      let newOrder=[...state.orders]
      newOrder.push( { objCustomerInfo: action.objCustomerInfo, products: action.objOrderInfo } );
      let newState={...state, orders: newOrder};

      return newState;
    }

              //обязатнльно пишем на тот случай если пришел action который не относится к этому reducery
    default:
      return state;
  }
}

export default orderReducer;