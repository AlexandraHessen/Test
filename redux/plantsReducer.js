// ---------------- САМА ФУНКЦИЯ REDUСER  ----------------//
// REDUCER не делает запрос в Ajax, логика должна быть в компоненте
import { PLANTS_LOADING, PLANTS_ERROR, PLANTS_SET } from './plantsAC'; //название action type


const initState={ // начальный state

  status: 0, // 0 - ничего не началось, 1 - идёт загрузка, 2 - была ошибка, 3 - данные загружены
  data: null, //даные которые мы загрузили по сети

}

function plantsReducer(state=initState,action) {
  switch (action.type) {

    case PLANTS_LOADING: {
      let newState={
        status:1,
        data:null,
      };
      return newState;
    }

    case PLANTS_ERROR: {
      let newState={
        status:2,
        data:null,
      };
      return newState;
    }

    case PLANTS_SET: {
      let newState={
        status:3,
        data:action.plants, //загруженные данные
      };
      return newState;
    }
    
//обязатнльно пишем на тот случай если пришел action который не относится к этому reducery
    default:
      return state;
  }
}

export default plantsReducer;
