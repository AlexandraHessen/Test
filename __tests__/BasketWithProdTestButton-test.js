"use strict";

import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';

import { createStore } from 'redux';
import combinedReducer from '../redux/reducers.js';
import { Provider } from 'react-redux';
import {add_product, clear_basket} from '../redux/basketAC';

import Basket  from '../components/Basket';
                    //импортируем файл который будем тестировать

let info = 	{
    "code": 1,
    "name": "Замиокулькас",
    "category": "deciduous",
    "price": 265,
    "description": "Полив по мере полного просыхания земляного кома, раз в 10-15 дней, в опрыскивании не нуждается. Чрезмерный полив замиокулькасов может приводить к пожелтению листьев.",
    "imgUrl": "/img/img_9241_1.jpg"
}

let store=createStore(combinedReducer);
store.dispatch(add_product(info.code,info));
expect(Object.keys(store.getState().basket.productsInBasket).length).toBe(1); 

test('BasketWithProdTestButton', () => {
                            // создаём тестовую версию компонента который будем тестировать
  const component = renderer.create(
  <Provider store={store}>
    <BrowserRouter>
      <Basket  basket = {info}/>
    </BrowserRouter>
  </Provider>

  );

                                // 1. получаем снэпшот (HTML-снимок) компонента для сверки, чтобы вёрстка не испортилась
    let componentTree=component.toJSON();
    expect(componentTree).toMatchSnapshot();
  
                              // найдём в вёрстке компонента саму кнопку
                              // в верстке <input type="button" value="Очистить корзину" className="BasketButtons" onClick = {this.clearBasket}></input>
                              const buttonAddProductToBasket = component.root.find( el => el.type=='input' && el.props.value == 'Очистить корзину'); 
                              // нажимаем на кнопку
      buttonAddProductToBasket.props.onClick();
  
                              // 2. получаем уже изменённый снэпшот
      componentTree=component.toJSON();
      expect(componentTree).toMatchSnapshot();
  
      store.dispatch(clear_basket());
      expect(Object.keys(store.getState().basket.productsInBasket).length).toBe(0); 

    
});

