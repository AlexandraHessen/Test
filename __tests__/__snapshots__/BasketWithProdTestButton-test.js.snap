// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasketWithProdTestButton 1`] = `
<div>
  <table
    className="TableBasket"
  >
    <tbody>
      <tr
        className="TableHeader"
      >
        <th>
          Товар
        </th>
        <th />
        <th
          className="priceTh"
        >
          Цена
        </th>
        <th />
      </tr>
      <tr
        className="RowProductBasket"
      >
        <td>
          <img
            className="Img"
            src="/img/img_9241_1.jpg"
          />
        </td>
        <td
          className="Name"
        >
          Замиокулькас
        </td>
        <td
          className="Price"
        >
          265
           руб.
        </td>
        <td>
          <button
            className="DelButton"
            onClick={[Function]}
            type="button"
          >
            <i
              className="far fa-times-circle"
            />
          </button>
        </td>
      </tr>
    </tbody>
  </table>
  <div
    className="Total"
  >
    Итого:
    <div
      className="SumProductsInBasket"
    >
      265
       руб.
    </div>
  </div>
  <div
    className="ActionsBasket"
  >
    <input
      className="BasketButtons"
      onClick={[Function]}
      type="button"
      value="Очистить корзину"
    />
    <input
      className="BasketButtons"
      onClick={[Function]}
      type="button"
      value="Оформить заказ"
    />
  </div>
</div>
`;

exports[`BasketWithProdTestButton 2`] = `
<div>
  <table
    className="TableBasket"
  >
    <tbody>
      <tr
        className="TableHeader"
      >
        <th>
          Товар
        </th>
        <th />
        <th
          className="priceTh"
        >
          Цена
        </th>
        <th />
      </tr>
    </tbody>
  </table>
  <div
    className="Total"
  >
    Итого:
    <div
      className="SumProductsInBasket"
    >
      0
       руб.
    </div>
  </div>
  <div
    className="ActionsBasket"
  >
    <input
      className="BasketButtons"
      onClick={[Function]}
      type="button"
      value="Очистить корзину"
    />
    <input
      className="BasketButtons"
      onClick={[Function]}
      type="button"
      value="Оформить заказ"
    />
  </div>
</div>
`;
